include/fiction/io/read_gate_level_layout.hpp

```diff
@@ -0,0 +1,349 @@
+ */
+class gate_level_parsing_error : public std::runtime_error
+{
+  public:
+    explicit gate_level_parsing_error(const std::string_view& msg) noexcept : std::runtime_error(msg.data()) {}
+};
+
+namespace detail
+{
+
+template <typename Lyt>
+class read_gate_level_layout_impl
+{
+  public:
+    read_gate_level_layout_impl(std::istream& s, const std::string_view& name) : lyt{}, is{s}
+    {
+        set_name(lyt, name);
+    }
+
+    read_gate_level_layout_impl(Lyt& tgt, std::istream& s) : lyt{tgt}, is{s} {}
+
+    Lyt run()
+    {
+        // tinyXML2 does not support std::istream, so we have to read the whole file into a string first
+        std::stringstream buffer{};
+        buffer << is.rdbuf();
+        const std::string sqd_content{buffer.str()};
+
+        tinyxml2::XMLDocument xml_document{};
+        xml_document.Parse(sqd_content.c_str());
+
+        if (xml_document.ErrorID() != 0)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: " + std::string(xml_document.ErrorName()));
+        }
+
+        auto* const fcn_root = xml_document.FirstChildElement("fcn");
+        if (fcn_root == nullptr)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: no root element 'fcn'");
+        }
+
+        auto* const layout = fcn_root->FirstChildElement("layout");
+        if (layout == nullptr)
+        {
+            throw gate_level_parsing_error("Error parsing gate_level file: no element 'layout'");
+        }
+
+        auto* const name = layout->FirstChildElement("name");
+        if (name != nullptr and name->GetText())
+        {
+            std::string layout_name = name->GetText();
+            set_name(lyt, layout_name);
+        }
+
+        auto* const clocking = layout->FirstChildElement("clocking");
+        if (clocking != nullptr)
+        {
+            lyt.replace_clocking_scheme(*get_clocking_scheme<Lyt>(clocking->GetText()));
+        }
+
+        auto* const size = layout->FirstChildElement("size");
+        if (size != nullptr)
+        {
+            int       x = std::stoi(size->FirstChildElement("x")->GetText());
+            int       y = std::stoi(size->FirstChildElement("y")->GetText());
+            int       z = std::stoi(size->FirstChildElement("z")->GetText());
+            tile<Lyt> max_pos{x, y, z};
+            lyt.resize(max_pos);
+        }
+
+        std::vector<gate_storage> gates;
+        auto* const               gates_xml = fcn_root->FirstChildElement("gates");
+        if (gates_xml != nullptr)
+        {
+            for (const auto* gate_xml = gates_xml->FirstChildElement("gate"); gate_xml != nullptr;
+                 gate_xml             = gate_xml->NextSiblingElement("gate"))
+            {
+                gate_storage gate = {};
+                gate.id           = std::stoi(gate_xml->FirstChildElement("id")->GetText());
+                gate.type         = gate_xml->FirstChildElement("type")->GetText();
```

Comment:
I'll stop mentioning it after this, but keep in mind that the fields you're parsing could be missing. Think about error handling.